File: AODashboard.Client/Model/Places.cs
// -----------------------------------------------------------------------
// <copyright file="Places.cs" company="Tony Richards">
// Copyright (c) Tony Richards. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
// </copyright>
// -----------------------------------------------------------------------

using System.Collections.Immutable;

namespace AODashboard.Client.Model;

/// <summary>
/// Represents a list of places.
/// </summary>
public readonly record struct Places
{
    /// <summary>
    /// Gets the list of place names.
    /// </summary>
    public IImmutableList<string> Names { get; init; }
}

File: AmbulanceDashboard.Client/src/api-hooks/places.ts
// -----------------------------------------------------------------------
// <copyright file="places.ts" company="Tony Richards">
// Copyright (c) Tony Richards. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for
// full license information.
// </copyright>
// -----------------------------------------------------------------------

import { useQuery } from "@tanstack/react-query";

export const useDistricts = (region: string) =>
  useQuery({
    queryKey: ["vehicles", "districts", region],
    queryFn: async () => {
      if (region === "All") {
        return [];
      }

      const response = await fetch(`/api/places/${region}/districts`);
      if (!response.ok) {
        throw new Error("Failed to fetch data");
      }

      const result = (await response.json()) as { names: string[] };

      return result.names;
    },
    throwOnError: true,
  });

export const useHubs = (region: string, district: string) =>
  useQuery({
    queryKey: ["vehicles", "hubs", region, district],
    queryFn: async () => {
      if (region === "All" || district === "All") {
        return [];
      }

      const response = await fetch(`/api/places/${region}/${district}/hubs`);
      if (!response.ok) {
        throw new Error("Failed to fetch data");
      }

      const result = (await response.json()) as { names: string[] };

      return result.names;
    },
    throwOnError: true,
  });
